tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c

```diff
@@ -229,6 +235,128 @@ static void test_rx_on_key_refresh(void)
 	PASS();
 }
 
+static void test_tx_beacon_init(void)
+{
+	bt_mesh_test_cfg_set(&tx_cfg, BEACON_OBSERVE_INTERVAL);
+}
+
+static void test_rx_beacon_init(void)
+{
+	bt_mesh_test_cfg_set(&tx_cfg, BEACON_OBSERVE_INTERVAL);
+}
+
+static void test_tx_beacon_secure(void)
+{
+	uint32_t delay_time;
+	uint32_t sent_beacons;
+	uint32_t sent_beacon_time;
+	struct bt_mesh_subnet *sub;
+	
+	/* shift beaconing time line to avoid boundary cases. */
+	delay_time = get_device_nbr() * 100;
+	k_sleep(K_MSEC(delay_time));
+
+	bt_mesh_test_setup();
+
+	sub = bt_mesh_subnet_find(NULL, NULL);
+
+	sent_beacons = 0;
+	sent_beacon_time = 0;
+	while (true) {
+		k_sleep(K_SECONDS(10));
+		if (sent_beacon_time != sub->beacon_sent) {
+			sent_beacon_time = sub->beacon_sent;
+			sent_beacons++;
+		}
+
+		if (sent_beacons >= 1) { /*sending out at least 1 beacon is enough*/
+			break;
+		}
+	}
+
+	PASS();
+}
+
+static struct k_sem observer_sem;
+static uint32_t obs_min_interval; /*minimum time measured between two consecutive beacons in milliseconds*/
+static uint32_t last_beacon_time; /*timestamp for last received valid secure beacon in milliseconds*/
+static uint32_t total_beacons; /*total number of beacons received*/
+
+static void scan_recv_cb(const struct bt_le_scan_recv_info *info, struct net_buf_simple *buf)
+{
+	struct bt_mesh_subnet *sub;
+	uint8_t length;
+	uint8_t payload[13];
+	uint64_t *sub_auth;
+	uint64_t auth;
+	uint8_t mac[16];
+	uint32_t now;
+	uint32_t interval;
+
+	now = k_uptime_get_32();
+	interval = now - last_beacon_time;
+
+	ASSERT_EQUAL(BT_GAP_ADV_TYPE_ADV_NONCONN_IND, info->adv_type);
+
+	length = net_buf_simple_pull_u8(buf);
+	ASSERT_EQUAL(buf->len, length);
+	ASSERT_EQUAL(BT_DATA_MESH_BEACON, net_buf_simple_pull_u8(buf));
+	ASSERT_EQUAL(0x01, net_buf_simple_pull_u8(buf));
+
+	sub = bt_mesh_subnet_find(NULL, NULL);
+
+	memcpy(payload, buf->data, sizeof(payload));
+	net_buf_simple_pull(buf, 13);
+
+	if (bt_mesh_aes_cmac_one(sub->keys->beacon, payload, sizeof(payload), mac)) {
+		return;
+	}
+
+	memcpy(&auth, mac, sizeof(auth));
+	sub_auth = (uint64_t *)sub->auth;
+
+	ASSERT_EQUAL(auth, *sub_auth);
+
+	last_beacon_time = now;
+
+	if (interval < obs_min_interval) {
+		obs_min_interval = interval;
+	}
+
+	total_beacons++;
+	if (total_beacons >= (EXPECTED_BEACON_CNT)) {
+		k_sem_give(&observer_sem);
+	}
+}
+
+static void scan_timeout_cb(void)
+{
+	return;
+}
+
+static struct bt_le_scan_cb scan_callbacks = {
+	.recv = scan_recv_cb,
+	.timeout = scan_timeout_cb,
+};
+
+static void test_rx_beacon_secure(void)
+{
+	k_sem_init(&observer_sem, 0, 1);
+	/*initial estimated interval value*/
+	obs_min_interval = 10000;
```

Comment:
If there were no beacons then the test will pass. I guess the initial value of `obs_min_interval` should be wrong in advance.

Reply:
good point, I should have chose a value just below the minimum.